src/VisualStudio/Core/Def/ProjectSystem/OpenTextBufferProvider.cs

```diff
@@ -0,0 +1,313 @@
+            {
+                // If we hadn't already raised an event for this, do it now
+                if (!_monikerToTextBufferMap.ContainsKey(moniker) && TryGetBufferFromRunningDocumentTable(docCookie, out var buffer))
+                {
+                    _monikerToTextBufferMap = _monikerToTextBufferMap.Add(moniker, buffer);
+                    _runningDocumentTable.GetDocumentHierarchyItem(docCookie, out var hierarchy, out _);
+
+                    RaiseEventForEachListener(l => l.OnOpenDocument(moniker, buffer, hierarchy));
+                }
+
+                RaiseEventForEachListener(l => l.OnDocumentOpenedIntoWindowFrame(moniker, pFrame));
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        public int OnAfterDocumentWindowHide(uint docCookie, IVsWindowFrame pFrame)
+            => VSConstants.E_NOTIMPL;
+
+        public int OnBeforeSave(uint docCookie)
+            => VSConstants.E_NOTIMPL;
+
+        /// <summary>
+        /// Attempts to get a text buffer from the specified file path. May be called on any thread.
+        /// </summary>
+        /// <param name="filePath">The file path to retrieve the text buffer for.</param>
+        /// <param name="textBuffer">The buffer if the file is open and initialized.</param>
+        /// <returns>true if the buffer was found with a non null value.</returns>
+        public bool TryGetBufferFromFilePath(string filePath, [NotNullWhen(true)] out ITextBuffer? textBuffer)
+        {
+            return _monikerToTextBufferMap.TryGetValue(filePath, out textBuffer);
+        }
+
+        public bool IsFileOpen(string filePath)
+        {
+            return _monikerToTextBufferMap.ContainsKey(filePath);
+        }
+
+        public IVsHierarchy? GetDocumentHierarchy(string filePath)
+        {
+            _foregroundAffinitization.AssertIsForeground();
+
+            if (!_runningDocumentTable.IsFileOpen(filePath))
+            {
+                return null;
+            }
+
+            var cookie = _runningDocumentTable.GetDocumentCookie(filePath);
+            _runningDocumentTable.GetDocumentHierarchyItem(cookie, out var hierarchy, out _);
+            return hierarchy;
+        }
+
+        /// <summary>
+        /// Enumerates the running document table to retrieve all initialized files. Must be called on the UI thread, since this returns <see cref="IVsHierarchy"/> objects.
+        /// </summary>
+        public IEnumerable<(string filePath, ITextBuffer textBuffer, IVsHierarchy hierarchy)> EnumerateDocumentSet()
+        {
+            _foregroundAffinitization.AssertIsForeground();
+
+            var documents = ArrayBuilder<(string, ITextBuffer, IVsHierarchy)>.GetInstance();
+            foreach (var cookie in GetInitializedRunningDocumentTableCookies())
+            {
+                if (TryGetMoniker(cookie, out var moniker) && TryGetBufferFromRunningDocumentTable(cookie, out var buffer))
```

Comment:
I'm wondering - should this be enumerating our map `_monikerToTextBufferMap`? 

Is there any concern that the textbuffer that is present in `_monikerToTextBufferMap` is different from the one we get back from the RDT?  I'm guessing not

Reply:
Since this is still fetching hierarchies which we don't hold onto, I decided to keep this as is. It's the existing code from the old implementation.